### PR TITLE
Smith/ratio bandits

### PR DESCRIPTION
**Pull request by lukebrawleysmith**
### Adding: 
1. ratio metrics
2. updated results for Bandits, including variation mean CIs, probabilities of best arm
3. updated flow for Exploratory results

### Testing

Added units tests for `get_bandit_response`.  

TODO:
1. explore why ratio metrics lag behind standalone metrics w/r/t best arm probabilities compared to A/B testing
2. add more unit tests

### CODE REVIEW

File: packages/back-end/src/services/stats.ts [hunk 1]
```diff
@@ -65,20 +65,25 @@ export interface AnalysisSettingsForStatsEngine {
...
-  decision_metric: string;
-  bandit_weights_seed: number;
   weights: { // todo: implement in python
     date: Date;
     weights?: number[];
   }[];
+  reweight: boolean; // todo: implement in python
+  decision_metric: string;
+  bandit_weights_seed: number;
+  alpha: number;
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/stats.ts:83]: use the analysis settings alpha? (org/exp setting)

File: packages/back-end/src/services/stats.ts [hunk 2]
```diff
@@ -65,20 +65,25 @@ export interface AnalysisSettingsForStatsEngine {
   phase_length_days: number;
   alpha: number;
   max_dimensions: number;
+  decision_metric: string;
+  bandit: boolean;
+  bandit_weights_seed: number;
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/stats.ts:70]: either move to BanditSettings... or move all BanditSettings to ANalysisSettings?

File: packages/back-end/types/experiment-snapshot.d.ts [hunk 3]
```diff
@@ -138,6 +138,7 @@ export interface SnapshotBanditSettings {
     date: Date;
     weights?: number[];
   }[];
+  alpha: number;
```
Comment (bryce-fitzsimons) [inline packages/back-end/types/experiment-snapshot.d.ts:141]: wondering if we should track this in ExperimentSnapshotAnalysisSettings? (although it only has consequence for bandits, I believe...)

File: packages/back-end/types/stats.d.ts [hunk 4]
```diff
@@ -7,6 +7,14 @@ export type PValueCorrection = null | "benjamini-hochberg" | "holm-bonferroni";
 export type DifferenceType = "relative" | "absolute" | "scaled";
 
 export type RiskType = "relative" | "absolute";
+
+interface BaseResponse {
```
Comment (bryce-fitzsimons) [inline packages/back-end/types/stats.d.ts:11]: what uses this?

File: packages/back-end/src/services/stats.ts (hunk 1 above)
Comment (bryce-fitzsimons) [reply packages/back-end/src/services/stats.ts:83]: remove this one

File: packages/back-end/types/experiment-snapshot.d.ts (hunk 3 above)
Comment (bryce-fitzsimons) [reply packages/back-end/types/experiment-snapshot.d.ts:141]: remove

File: packages/back-end/types/stats.d.ts (hunk 4 above)
Comment (bryce-fitzsimons) [reply packages/back-end/types/stats.d.ts:11]: remove

File: packages/back-end/types/stats.d.ts [hunk 5]
```diff
@@ -60,15 +68,24 @@ export type ExperimentMetricAnalysis = {
...
+  ci?: [number, number];
+};
+
+export type BanditResult = {
+  singleVariationResults?: SingleVariationResult[];
   banditWeights?: number[];
+  bestArmProbabilities?: number[];
   additionalReward: number;
+  seed: number;
+  banditUpdateMessage?: string;
```
Comment (bryce-fitzsimons) [inline packages/back-end/types/stats.d.ts:83]: `banditError?: string;` for anything that breaks the bandit

File: packages/back-end/src/services/stats.ts (hunk 1 above)
Comment (lukebrawleysmith) [reply packages/back-end/src/services/stats.ts:83]: now using analysis settings alpha and removed alpha from BanditSettingsForStatsEngine

File: packages/back-end/src/services/stats.ts (hunk 2 above)
Comment (lukebrawleysmith) [reply packages/back-end/src/services/stats.ts:70]: moved to `BanditSettingsForStatsEngine`

File: packages/back-end/types/experiment-snapshot.d.ts (hunk 3 above)
Comment (lukebrawleysmith) [reply packages/back-end/types/experiment-snapshot.d.ts:141]: removed

File: packages/back-end/types/stats.d.ts (hunk 4 above)
Comment (lukebrawleysmith) [reply packages/back-end/types/stats.d.ts:11]: removed

File: packages/back-end/types/stats.d.ts (hunk 5 above)
Comment (lukebrawleysmith) [reply packages/back-end/types/stats.d.ts:83]: added `banditError?: string;`

**Pull request merged**